pkg/manager/manager.go

```diff
@@ -0,0 +1,167 @@
+	"sigs.k8s.io/controller-runtime/pkg/recorder"
+)
+
+// Manager initializes shared dependencies such as Caches and Clients, and provides them to runnables.
+type Manager interface {
+	// Add will set reqeusted dependencies on the component, and cause the component to be
+	// started when Start is called.  Add will inject any dependencies for which the argument
+	// implements the inject interface - e.g. inject.Client
+	Add(Runnable) error
+
+	// SetFields will set any dependencies on an object for which the object has implemented the inject
+	// interface - e.g. inject.Client.
+	SetFields(interface{}) error
+
+	// Start starts all registered Controllers and blocks until the Stop channel is closed.
+	// Returns an error if there is an error starting any controller.
+	Start(<-chan struct{}) error
+
+	// GetConfig returns an initialized Config
+	GetConfig() *rest.Config
+
+	// GetScheme returns and initialized Scheme
+	GetScheme() *runtime.Scheme
+
+	// GetClient returns a client configured with the Config
+	GetClient() client.Client
+
+	// GetFieldIndexer returns a client.FieldIndexer configured with the client
+	GetFieldIndexer() client.FieldIndexer
+
+	// GetCache returns a cache.Cache
+	GetCache() cache.Cache
+
+	// GetRecorder returns a new EventRecorder for the provided name
+	GetRecorder(name string) record.EventRecorder
+}
+
+// Options are the arguments for creating a new Manager
+type Options struct {
+	// Scheme is the scheme used to resolve runtime.Objects to GroupVersionKinds / Resources
+	// Defaults to the kubernetes/client-go scheme.Scheme
+	Scheme *runtime.Scheme
+
+	// Mapper is the rest mapper used to map go types to Kubernetes APIs
+	MapperProvider func(c *rest.Config) (meta.RESTMapper, error)
+
+	// Dependency injection for testing
+	newCache            func(config *rest.Config, opts cache.Options) (cache.Cache, error)
+	newClient           func(config *rest.Config, options client.Options) (client.Client, error)
+	newRecorderProvider func(config *rest.Config, scheme *runtime.Scheme) (recorder.Provider, error)
+}
+
+// Runnable allows a component to be started.
+type Runnable interface {
+	// Start starts running the component.  The component will stop running when the channel is closed.
+	// Start blocks until the channel is closed or an error occurs.
+	Start(<-chan struct{}) error
+}
+
+// RunnableFunc implements Runnable
+type RunnableFunc func(<-chan struct{}) error
+
+// Start implements Runnable
+func (r RunnableFunc) Start(s <-chan struct{}) error {
+	return r(s)
+}
+
+// New returns a new Manager
+func New(config *rest.Config, options Options) (Manager, error) {
+	cm := &controllerManager{config: config, scheme: options.Scheme, errChan: make(chan error)}
+
+	// Initialize a rest.config if none was specified
+	if cm.config == nil {
+		return nil, fmt.Errorf("must specify Config")
+	}
+
+	// Use the Kubernetes client-go scheme if none is specified
+	if cm.scheme == nil {
+		cm.scheme = scheme.Scheme
+	}
```

Comment:
This should be moved to `setOptionsDefaults`

Reply:
Done.  Restructured this fn.